File: sorting/merge_sort.c
#include <stdio.h>
#include <stdlib.h>

// Merges two sorted subarrays A[l..m] and A[m+1..r]
void merge(int A[], int l, int m, int r)
{
    int i = l;
    int j = m + 1;
    int k = l;

    // Temporary array to store merged result
    int B[100]; // Assuming array size won't exceed 100

    while (i <= m && j <= r)
    {
        if (A[i] <= A[j])
        {
            B[k] = A[i];
            i++;
        }
        else
        {
            B[k] = A[j];
            j++;
        }
        k++;
    }

    // Copy remaining elements from left subarray
    while (i <= m)
    {
        B[k] = A[i];
        i++;
        k++;
    }

    // Copy remaining elements from right subarray
    while (j <= r)
    {
        B[k] = A[j];
        j++;
        k++;
    }

    // Copy merged elements back to original array
    for (int p = l; p <= r; p++)
    {
        A[p] = B[p];
    }
}

void mergeSort(int A[], int l, int r)
{
    if (l < r)
    {
        int m = (l + r) / 2;
        mergeSort(A, l, m);
        mergeSort(A, m + 1, r);
        merge(A, l, m, r);
    }
}

int main()
{
    int A[] = {38, 27, 43, 3, 9, 82, 10};
    int n = sizeof(A) / sizeof(A[0]);

    printf("Original array:\n");
    for (int i = 0; i < n; i++)
        printf("%d ", A[i]);
    printf("\n");

    mergeSort(A, 0, n - 1);

    printf("Sorted array:\n");
    for (int i = 0; i < n; i++)
        printf("%d ", A[i]);
    printf("\n");

    return 0;
}

File: sorting/merge_sort.py
def merge(arr, l, m, r):
    left = arr[l:m+1]
    right = arr[m+1:r+1]

    i = j = 0  
    k = l      

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            arr[k] = left[i]
            i += 1
        else:
            arr[k] = right[j]
            j += 1
        k += 1

    while i < len(left):
        arr[k] = left[i]
        i += 1
        k += 1

    while j < len(right):
        arr[k] = right[j]
        j += 1
        k += 1

def merge_sort(arr, l, r):
    if l < r:
        m = (l + r) // 2
        merge_sort(arr, l, m)  #sort left half
        merge_sort(arr, m + 1, r) #sort right half
        merge(arr, l, m, r) #Merge the halves

arr = [38, 27, 43, 3, 9, 82, 10]
print("Original array:", arr)
merge_sort(arr, 0, len(arr) - 1)
print("Sorted array:", arr)
